### ISSUE TITLE
Suggestion: More helpful error message on VASP job crashes than VasprunXMLValidator error

### ISSUE THREAD

**Issue by Andrew-S-Rosen** [materialsproject/custodian#217]
Currently, when a VASP job fails for some machine-related reason (e.g. SLURM error), the traceback points to a `VasprunXMLValidator` error like that shown below. Especially for new users, this is not particularly insightful and can actually lead to some degree of confusion because the user might think there is something wrong with parsing of the `vasprun.xml` file when, in fact, it's a more fundamental error.

Before checking the validity of the `vasprun.xml` file, we should have some quick sanity checks to throw a more helpful error message if it's clear the job didn't run or if this is some more fundamental issue.

```
    run_custodian()
  File "/global/homes/r/rosen/software/miniconda/envs/cms/lib/python3.10/site-packages/quacc/custodian/vasp.py", line 136, in run_custodian
    c.run()
  File "/global/homes/r/rosen/software/miniconda/envs/cms/lib/python3.10/site-packages/custodian/custodian.py", line 367, in run
    self._run_job(job_n, job)
  File "/global/homes/r/rosen/software/miniconda/envs/cms/lib/python3.10/site-packages/custodian/custodian.py", line 489, in _run_job
    raise ValidationError(s, True, v)
custodian.custodian.ValidationError: Validation failed: VasprunXMLValidator
Traceback (most recent call last):
  File "/global/homes/r/rosen/software/miniconda/envs/cms/lib/python3.10/site-packages/fireworks/core/rocket.py", line 261, in run
    m_action = t.run_task(my_spec)
  File "/global/homes/r/rosen/software/miniconda/envs/cms/lib/python3.10/site-packages/jobflow/managers/fireworks.py", line 160, in run_task
    response = job.run(store=store)
  File "/global/homes/r/rosen/software/miniconda/envs/cms/lib/python3.10/site-packages/jobflow/core/job.py", line 530, in run
    response = function(*self.function_args, **self.function_kwargs)
  File "/global/homes/r/rosen/software/miniconda/envs/cms/lib/python3.10/site-packages/quacc/recipes/vasp/qmof.py", line 71, in make
    summary1 = _prerelax(atoms, self.preset, swaps, fmax=5.0)
  File "/global/homes/r/rosen/software/miniconda/envs/cms/lib/python3.10/site-packages/quacc/recipes/vasp/qmof.py", line 150, in _prerelax
    dyn.run(fmax=fmax)
  File "/global/homes/r/rosen/software/miniconda/envs/cms/lib/python3.10/site-packages/ase/optimize/optimize.py", line 273, in run
    return Dynamics.run(self)
  File "/global/homes/r/rosen/software/miniconda/envs/cms/lib/python3.10/site-packages/ase/optimize/optimize.py", line 156, in run
    for converged in Dynamics.irun(self):
  File "/global/homes/r/rosen/software/miniconda/envs/cms/lib/python3.10/site-packages/ase/optimize/optimize.py", line 135, in irun
    self.step()
  File "/global/homes/r/rosen/software/miniconda/envs/cms/lib/python3.10/site-packages/ase/optimize/bfgslinesearch.py", line 122, in step
    ls._line_search(self.func, self.fprime, r, self.p, g, e, self.e0,
  File "/global/homes/r/rosen/software/miniconda/envs/cms/lib/python3.10/site-packages/ase/utils/linesearch.py", line 57, in _line_search
    fval = func(xk + stp * pk, *args)
  File "/global/homes/r/rosen/software/miniconda/envs/cms/lib/python3.10/site-packages/ase/optimize/bfgslinesearch.py", line 171, in func
    return (self.atoms.get_potential_energy(
  File "/global/homes/r/rosen/software/miniconda/envs/cms/lib/python3.10/site-packages/ase/atoms.py", line 728, in get_potential_energy
    energy = self._calc.get_potential_energy(
  File "/global/homes/r/rosen/software/miniconda/envs/cms/lib/python3.10/site-packages/ase/calculators/abc.py", line 24, in get_potential_energy
    return self.get_property(name, atoms)
  File "/global/homes/r/rosen/software/miniconda/envs/cms/lib/python3.10/site-packages/ase/calculators/calculator.py", line 509, in get_property
    self.calculate(atoms, [name], system_changes)
  File "/global/homes/r/rosen/software/miniconda/envs/cms/lib/python3.10/site-packages/ase/calculators/vasp/vasp.py", line 325, in calculate
    raise calculator.CalculationFailed(
ase.calculators.calculator.CalculationFailed: vasp in . returned an error: 1
```

**Comment by rkingsbury** [comment]
100% agree; this cryptic catch-all error is _the_ biggest obstacle to efficient troubleshooting imho.  As far as improving the situation, as Kristin would say "the devil is in the details". 

I know the proximate cause of this issue is that a failed VASP calculation won't properly close the `</xml>` tags, making it impossible to parse the file as XML. One thing we could do is insert a try/except along the lines of:

```
try:
    <parse vasprun.xml>
except XMLError:
    ## Read the .xml file as text, line by line, searching for specific keys to indicate what happened
    for line in vasprun.readlines():
         if <key1> in line:
             <do something?
```

The other suggestion I have is if it's possible to parse `vasp.out` _prior to_ `vasprun.xml`, we could search for certain patterns like out of memory and such.

I'm not terribly familiar with the technical ins and outs of all this so the above are about as deep as my suggestions will go. Props for raising this issue though; if we could handle it better it would greatly improve the usability of our infrastructure for VASP.

**Comment by janosh** [comment]
Perhaps worth mentioning that [BeautifulSoup](https://www.crummy.com/software/BeautifulSoup/bs4/doc) has very good parsing support for incomplete XML files.

**Comment by mkhorton** [comment]
> Perhaps worth mentioning that [BeautifulSoup](https://www.crummy.com/software/BeautifulSoup/bs4/doc) has very good parsing support for incomplete XML files.

The file being incomplete in itself is rarely helpful though (e.g., even parsing the incomplete vasprun.xml would probably give us little extra useful info, the issue is that the job was stopped prematurely). 

> The other suggestion I have is if it's possible to parse vasp.out prior to vasprun.xml, we could search for certain patterns like out of memory and such.

Yes, or std_error too, etc. If you're running via FireWorks there's already the ability to add "trackers" for these which can save some time, but isn't a particularly good solution overall.

**Comment by Andrew-S-Rosen** [comment]
It's my understanding that this can finally (?) be closed as of #264!
